config/config.go

```diff
@@ -36,7 +36,12 @@ const (
 	MinNumFiles = 1
 )
 
-var DefaultDataDir = filepath.Join(smutil.GetUserHomeDirectory(), "post", DefaultDataDirName)
+var DefaultDataDir string
+
+func init() {
+	home, _ := os.UserHomeDir()
```

Comment:
I intentionally didn't, the code I replaced returned `""` if it was unable to determine the users home directory without panicing. Since this is what `os.UserHomeDir()` does too in the case of an error I just chose to ignore the `error`.

Reply:
I'm not sure if the replaced code was right about this. According to the [docs](https://pkg.go.dev/os#UserHomeDir):
> On Unix, including macOS, it returns the $HOME environment variable. On Windows, it returns %USERPROFILE%. On Plan 9, it returns the $home environment variable.

Is just using `""` actually **OK**? It might end up treating the current directory as "home". This will lead to messing up whatever working directory is at the moment. I find it wrong and dangerous. IMO, this is an unexpected situation and should be handled somehow (a panic, a warning log + using /tmp/ etc).

Reply:
Imho raising a panic prevents a usecase that might be important to some:

A smesher decides to run go-spacemesh as daemon with a user that is only used for that process and doesn't have a home directory, password, configured shell, etc. (as it is customary for daemons). The user sets the datadir explicitly for go-spacemesh so it writes to `/opt/spacemesh/` (or something similar). If we add an error check here the program stops not because the data directory is invalid, but because the **default** dataDir couldn't be determined. Logging in this case would also give not valuable information (why is not being able to determine a path that isn't used noteworthy to be added to the logs?)

If the smesher forgets to set a dataDir it would default to `/post/data` (or the equivalent on non-linux systems). If they are running it with root permissions then go-spacemesh just writes there (we don't babysit admins :wink:) if they don't give it root permissions it will fail to create the directory or access files within and tell them so: The first time probably when go-spacemesh tries to load / create the identify file and isn't able to.

If we default to `/tmp` this might cause other issues: go-spacemesh might be running on a system that doesn't have a `/tmp`. `panic` would prevent the user from starting the application at all even if they explicitly set the DataDir.